awscli/clidocs.py

```diff
@@ -186,18 +190,47 @@ def doc_relateditem(self, help_command, related_item, **kwargs):
         )
         doc.write('\n')
 
-    def _document_enums(self, argument, doc):
+    def _document_enums(self, model, doc):
         """Documents top-level parameter enums"""
-        if hasattr(argument, 'argument_model'):
-            model = argument.argument_model
-            if isinstance(model, StringShape):
-                if model.enum:
-                    doc.style.new_paragraph()
-                    doc.write('Possible values:')
-                    doc.style.start_ul()
-                    for enum in model.enum:
-                        doc.style.li('``%s``' % enum)
-                    doc.style.end_ul()
+        if isinstance(model, StringShape):
+            if model.enum:
+                doc.style.new_paragraph()
+                doc.write('Possible values:')
+                doc.style.start_ul()
+                for enum in model.enum:
+                    doc.style.li('``%s``' % enum)
+                doc.style.end_ul()
+
+    def _document_nested_structure(self, model, doc, stack):
+        """Recursively documents parameters in nested structures"""
+        shape_name = getattr(model, 'name', None)
+        if not shape_name or shape_name in stack:
+            return
+        stack.append(shape_name)
+        if isinstance(model, StructureShape):
+            self._output_struct_data(model.members, doc, stack)
+        if isinstance(model, ListShape):
+            self._output_list_data(model, doc, stack)
+        stack.pop()
+
+    def _output_struct_data(self, members, doc, stack):
+        doc.style.start_ul()
+        for name, shape in members.items():
+            doc.style.start_li()
+            doc.write('**%s** (%s) -- ' % (name, shape.type_name))
+            doc.parser.feed(shape.documentation)
+            self._document_nested_structure(shape, doc, stack)
+            doc.style.end_li()
+        doc.style.end_ul()
+
+    def _output_list_data(self, data, doc, stack):
+        doc.style.start_ul()
+        doc.style.start_li()
+        doc.write('(%s) ' % data.member.type_name)
+        doc.parser.feed(data.member.documentation)
+        self._document_nested_structure(data.member, doc, stack)
+        doc.style.end_li()
+        doc.style.end_ul()
```

Comment:
We should add tests to make sure we are correctly documenting nested input structures. I think we can added cases to this test file: https://github.com/aws/aws-cli/blob/develop/tests/unit/test_clidocs.py and add cases that the expected documentation is produced at least for (there may be more):
1. Nested structures
2. Nested maps
3. Nested lists
4. We handle shapes that can result in hitting circular recursions.